File: plugins/plugin-memory/python/elizaos_plugin_memory/actions/forget.py
"""FORGET action - remove a stored memory by ID or by matching content."""

from __future__ import annotations

import json
import logging
import re

from elizaos_plugin_memory.actions.base import (
    ActionExample,
    ActionResult,
    HandlerCallback,
    Memory,
    RuntimeProtocol,
    State,
    create_action,
)
from elizaos_plugin_memory.types import (
    MEMORY_SOURCE,
    PLUGIN_MEMORY_TABLE,
    decode_memory_text,
)

logger = logging.getLogger(__name__)


async def validate(
    runtime: RuntimeProtocol,
    _message: Memory,
    _state: State | None = None,
) -> bool:
    try:
        return callable(getattr(runtime, "get_memories", None)) and callable(
            getattr(runtime, "delete_memory", None)
        )
    except Exception:
        return False


async def handler(
    runtime: RuntimeProtocol,
    message: Memory,
    _state: State | None = None,
    options: dict[str, str | list[str] | int | None] | None = None,
    callback: HandlerCallback | None = None,
) -> ActionResult:
    try:
        content_data = message.get("content", {})
        content = content_data.get("text", "")
        if not content:
            error_message = "Please specify which memory to forget."
            if callback:
                await callback({"text": error_message, "source": content_data.get("source")})
            return {"text": error_message, "success": False}

        # Direct removal by ID
        memory_id = str(options.get("memoryId", "")) if options else ""
        if memory_id:
            await runtime.delete_memory(memory_id)
            success_message = f"Removed memory with ID: {memory_id}"
            if callback:
                await callback({"text": success_message, "source": content_data.get("source")})
            return {"text": success_message, "success": True, "data": {"removedId": memory_id}}

        # Search for matching memory
        memories = await runtime.get_memories(
            {
                "roomId": message.get("roomId", ""),
                "tableName": PLUGIN_MEMORY_TABLE,
                "count": 100,
            }
        )

        plugin_memories = [
            m for m in memories if m.get("content", {}).get("source") == MEMORY_SOURCE
        ]

        if not plugin_memories:
            no_mem_msg = "No stored memories found to remove."
            if callback:
                await callback({"text": no_mem_msg, "source": content_data.get("source")})
            return {"text": no_mem_msg, "success": True}

        # Use LLM to find the best matching memory
        search_content = str(options.get("content", content)) if options else content
        descriptions = []
        for i, m in enumerate(plugin_memories):
            parsed = decode_memory_text(m.get("content", {}).get("text", ""))
            descriptions.append(f'{i}: "{parsed.content}"')

        match_prompt = (
            "Given the user's request to forget a memory, identify which memory index matches.\n\n"
            f'User request: "{search_content}"\n\n'
            f"Available memories:\n{chr(10).join(descriptions)}\n\n"
            'Return ONLY a JSON object: {"index": <number or -1>, "confidence": <0.0 to 1.0>}'
        )

        response = await runtime.use_model("TEXT_LARGE", {"prompt": match_prompt})
        if not response:
            raise RuntimeError("Failed to identify memory to remove")

        cleaned = re.sub(r"^```(?:json)?\n?", "", response)
        cleaned = re.sub(r"\n?```$", "", cleaned).strip()
        match_result: dict[str, int | float] = json.loads(cleaned)

        idx = int(match_result.get("index", -1))
        confidence = float(match_result.get("confidence", 0.0))

        if idx < 0 or idx >= len(plugin_memories) or confidence < 0.5:
            no_match_msg = "Could not find a matching memory to remove. Please be more specific."
            if callback:
                await callback({"text": no_match_msg, "source": content_data.get("source")})
            return {"text": no_match_msg, "success": False}

        target = plugin_memories[idx]
        parsed = decode_memory_text(target.get("content", {}).get("text", ""))
        target_id = target.get("id", "")

        if target_id:
            await runtime.delete_memory(target_id)

        success_message = f'Removed memory: "{parsed.content}"'
        if callback:
            await callback({"text": success_message, "source": content_data.get("source")})

        return {
            "text": success_message,
            "success": True,
            "data": {"removedId": target_id, "content": parsed.content},
        }

    except Exception as error:
        logger.error("Failed to forget memory: %s", error)
        error_message = f"Failed to forget memory: {error}"
        if callback:
            await callback(
                {"text": error_message, "source": message.get("content", {}).get("source")}
            )
        return {"text": error_message, "success": False}


forget_action = create_action(
    name="FORGET",
    description="Remove a stored memory by ID or by matching content description",
    similes=["forget", "remove-memory", "delete-memory", "erase-memory", "clear-memory"],
    examples=[
        [
            ActionExample(
                name="User",
                content={"text": "Forget what you know about my favorite color."},
            ),
            ActionExample(
                name="Assistant",
                content={
                    "text": "I'll remove that memory about your favorite color.",
                    "actions": ["FORGET"],
                },
            ),
        ],
        [
            ActionExample(
                name="User",
                content={"text": "Delete the memory about the project deadline."},
            ),
            ActionExample(
                name="Assistant",
                content={
                    "text": "I've removed the memory about the project deadline.",
                    "actions": ["FORGET"],
                },
            ),
        ],
    ],
    validate=validate,
    handler=handler,
)

File: plugins/plugin-memory/typescript/src/actions/forget.ts
import {
  type Action,
  type ActionResult,
  type HandlerCallback,
  type HandlerOptions,
  type IAgentRuntime,
  logger,
  type Memory,
  ModelType,
  type State,
} from "@elizaos/core";
import {
  decodeMemoryText,
  type ForgetParameters,
  MEMORY_SOURCE,
  PLUGIN_MEMORY_TABLE,
} from "../types.js";

export const forgetAction: Action = {
  name: "FORGET",
  description: "Remove a stored memory by ID or by matching content description",
  similes: ["forget", "remove-memory", "delete-memory", "erase-memory", "clear-memory"],

  examples: [
    [
      {
        name: "User",
        content: { text: "Forget what you know about my favorite color." },
      },
      {
        name: "Assistant",
        content: {
          text: "I'll remove that memory about your favorite color.",
          actions: ["FORGET"],
        },
      },
    ],
    [
      {
        name: "User",
        content: { text: "Delete the memory about the project deadline." },
      },
      {
        name: "Assistant",
        content: {
          text: "I've removed the memory about the project deadline.",
          actions: ["FORGET"],
        },
      },
    ],
  ],

  async validate(runtime: IAgentRuntime, _message: Memory, _state?: State): Promise<boolean> {
    return typeof runtime.getMemories === "function" && typeof runtime.deleteMemory === "function";
  },

  async handler(
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
    _options?: HandlerOptions,
    callback?: HandlerCallback
  ): Promise<ActionResult> {
    try {
      const content = message.content.text;
      if (!content) {
        const errorMessage = "Please specify which memory to forget.";
        await callback?.({ text: errorMessage, source: message.content.source });
        return { text: errorMessage, success: false };
      }

      const params = _options?.parameters as ForgetParameters | undefined;

      // Direct removal by ID if provided
      if (params?.memoryId) {
        await runtime.deleteMemory(params.memoryId as import("@elizaos/core").UUID);
        const successMessage = `Removed memory with ID: ${params.memoryId}`;
        await callback?.({ text: successMessage, source: message.content.source });
        return { text: successMessage, success: true, data: { removedId: params.memoryId } };
      }

      // Search for matching memory to remove
      const memories = await runtime.getMemories({
        roomId: message.roomId,
        tableName: PLUGIN_MEMORY_TABLE,
        count: 100,
      });

      const pluginMemories = memories.filter((m) => m.content.source === MEMORY_SOURCE);

      if (pluginMemories.length === 0) {
        const noMemoriesMsg = "No stored memories found to remove.";
        await callback?.({ text: noMemoriesMsg, source: message.content.source });
        return { text: noMemoriesMsg, success: true };
      }

      // Use LLM to identify which memory to forget
      const searchContent = params?.content ?? content;
      const memoryDescriptions = pluginMemories.map((m, i) => {
        const parsed = decodeMemoryText(m.content.text);
        return `${i}: "${parsed.content}"`;
      });

      const matchPrompt = `Given the user's request to forget a memory, identify which stored memory index matches best.

User request: "${searchContent}"

Available memories:
${memoryDescriptions.join("\n")}

Return ONLY a JSON object (no markdown, no code blocks):
{"index": <number or -1 if no match>, "confidence": <0.0 to 1.0>}`;

      const response = await runtime.useModel(ModelType.TEXT_LARGE, {
        prompt: matchPrompt,
      });

      if (!response) {
        throw new Error("Failed to identify memory to remove");
      }

      const cleaned = response
        .replace(/^```(?:json)?\n?/, "")
        .replace(/\n?```$/, "")
        .trim();
      const match: { index: number; confidence: number } = JSON.parse(cleaned);

      if (match.index < 0 || match.index >= pluginMemories.length || match.confidence < 0.5) {
        const noMatchMsg = "Could not find a matching memory to remove. Please be more specific.";
        await callback?.({ text: noMatchMsg, source: message.content.source });
        return { text: noMatchMsg, success: false };
      }

      const targetMemory = pluginMemories[match.index];
      const parsed = decodeMemoryText(targetMemory.content.text);
      const memoryId = targetMemory.id ?? "";

      if (memoryId) {
        await runtime.deleteMemory(memoryId as import("@elizaos/core").UUID);
      }

      const successMessage = `Removed memory: "${parsed.content}"`;
      await callback?.({ text: successMessage, source: message.content.source });

      return {
        text: successMessage,
        success: true,
        data: { removedId: memoryId, content: parsed.content },
      };
    } catch (error) {
      logger.error("Failed to forget memory:", error);
      const errorMessage = `Failed to forget memory: ${error instanceof Error ? error.message : String(error)}`;
      await callback?.({ text: errorMessage, source: message.content.source });
      return { text: errorMessage, success: false };
    }
  },
};
